File: data_baseline/openrouter__qwen__qwen3.7-plus/haskell/rep8/Main.hs
{-# LANGUAGE OverloadedStrings #-}

module Main where

import Control.Concurrent.STM (TVar, newTVarIO, readTVarIO, atomically, writeTVar)
import Control.Monad.IO.Class (liftIO)
import Data.Aeson (FromJSON, ToJSON, Value, object, (.=), (.:?), withObject, parseJSON, eitherDecode)
import qualified Data.Aeson as Aeson
import Data.Char (isAsciiLower, isAsciiUpper, isDigit)
import Data.Function (on)
import qualified Data.List as L
import qualified Data.Map as M
import Data.Maybe (fromMaybe)
import Data.Text (Text)
import qualified Data.Text as T
import qualified Data.Text.Lazy as TL
import qualified Data.Text.Encoding as TE
import qualified Data.ByteString.Lazy.Char8 as BSL
import Data.Time (getCurrentTime, formatTime, defaultTimeLocale)
import Network.HTTP.Types (status200, status201, status204, status400, status401, status404, status409, status500)
import Network.Wai (rawPathInfo)
import System.Environment (getArgs)
import System.Random (randomIO)
import Text.Read (readMaybe)
import Web.Scotty (request, get, post, put, delete, scotty, ActionM, json, status, finish, body, getCookie, setHeader)

data User = User
  { userId :: Int
  , username :: Text
  , password :: Text
  } deriving (Show)

data Todo = Todo
  { todoId :: Int
  , todoUserId :: Int
  , todoTitle :: Text
  , todoDescription :: Text
  , todoCompleted :: Bool
  , todoCreatedAt :: Text
  , todoUpdatedAt :: Text
  } deriving (Show)

data AppState = AppState
  { nextUserId :: Int
  , users :: M.Map Int User
  , usernameToId :: M.Map Text Int
  , nextTodoId :: Int
  , todos :: M.Map Int Todo
  , sessions :: M.Map Text Int
  } deriving (Show)

initialState :: AppState
initialState = AppState
  { nextUserId = 1
  , users = M.empty
  , usernameToId = M.empty
  , nextTodoId = 1
  , todos = M.empty
  , sessions = M.empty
  }

instance ToJSON User where
  toJSON u = object
    [ "id" .= userId u
    , "username" .= username u
    ]

instance ToJSON Todo where
  toJSON t = object
    [ "id" .= todoId t
    , "title" .= todoTitle t
    , "description" .= todoDescription t
    , "completed" .= todoCompleted t
    , "created_at" .= todoCreatedAt t
    , "updated_at" .= todoUpdatedAt t
    ]

data RegisterReq = RegisterReq { regUsername :: Maybe Text, regPassword :: Maybe Text }
instance FromJSON RegisterReq where
  parseJSON = withObject "RegisterReq" $ \v -> RegisterReq
    <$> v .:? "username"
    <*> v .:? "password"

data LoginReq = LoginReq { loginUsername :: Maybe Text, loginPassword :: Maybe Text }
instance FromJSON LoginReq where
  parseJSON = withObject "LoginReq" $ \v -> LoginReq
    <$> v .:? "username"
    <*> v .:? "password"

data PasswordReq = PasswordReq { oldPassword :: Maybe Text, newPassword :: Maybe Text }
instance FromJSON PasswordReq where
  parseJSON = withObject "PasswordReq" $ \v -> PasswordReq
    <$> v .:? "old_password"
    <*> v .:? "new_password"

data TodoCreateReq = TodoCreateReq { tcTitle :: Maybe Text, tcDescription :: Maybe Text }
instance FromJSON TodoCreateReq where
  parseJSON = withObject "TodoCreateReq" $ \v -> TodoCreateReq
    <$> v .:? "title"
    <*> v .:? "description"

data TodoUpdateReq = TodoUpdateReq
  { tuTitle :: Maybe Text
  , tuDescription :: Maybe Text
  , tuCompleted :: Maybe Bool
  }
instance FromJSON TodoUpdateReq where
  parseJSON = withObject "TodoUpdateReq" $ \v -> TodoUpdateReq
    <$> v .:? "title"
    <*> v .:? "description"
    <*> v .:? "completed"

isValidUsername :: Text -> Bool
isValidUsername t =
  let len = T.length t
      validChars = T.all (\c -> isAsciiLower c || isAsciiUpper c || isDigit c || c == '_') t
  in len >= 3 && len <= 50 && validChars

generateToken :: IO Text
generateToken = do
  chars <- mapM (\_ -> do
                   i <- randomIO :: IO Int
                   let hexChars = "0123456789abcdef"
                   return $ hexChars !! (i `mod` 16)
                ) [1..32]
  return $ T.pack chars

getCurrentTimeStr :: IO Text
getCurrentTimeStr = do
  now <- getCurrentTime
  return $ T.pack $ formatTime defaultTimeLocale "%Y-%m-%dT%H:%M:%SZ" now

jsonData' :: FromJSON a => ActionM a
jsonData' = do
  b <- body
  case eitherDecode b of
    Left _ -> do
      status status400
      json (Aeson.object ["error" .= ("Invalid JSON" :: Text)] :: Aeson.Value)
      finish
    Right val -> return val

requireAuth :: TVar AppState -> ActionM Int
requireAuth stateTVar = do
  mSession <- getCookie "session_id"
  case mSession of
    Nothing -> do
      status status401
      json (Aeson.object ["error" .= ("Authentication required" :: Text)] :: Aeson.Value)
      finish
    Just token -> do
      st <- liftIO $ readTVarIO stateTVar
      case M.lookup token (sessions st) of
        Nothing -> do
          status status401
          json (Aeson.object ["error" .= ("Authentication required" :: Text)] :: Aeson.Value)
          finish
        Just uid -> return uid

getPortArg :: IO Int
getPortArg = do
  args <- getArgs
  case args of
    ["--port", p] -> return (read p)
    _ -> return 3000

getPathParam :: ActionM Text
getPathParam = do
  req <- request
  let path = TE.decodeUtf8 (rawPathInfo req)
      segments = T.split (== '/') path
  return $ if length segments >= 3 then segments !! 2 else ""

main :: IO ()
main = do
  portArg <- getPortArg
  state <- newTVarIO initialState
  scotty portArg $ do
    
    post "/register" $ do
      req <- jsonData' :: ActionM RegisterReq
      let mU = regUsername req
      let mP = regPassword req
      case mU of
        Nothing -> do
          status status400
          json (Aeson.object ["error" .= ("Invalid username" :: Text)] :: Aeson.Value)
        Just u ->
          if not (isValidUsername u)
            then do
              status status400
              json (Aeson.object ["error" .= ("Invalid username" :: Text)] :: Aeson.Value)
            else case mP of
              Nothing -> do
                status status400
                json (Aeson.object ["error" .= ("Password too short" :: Text)] :: Aeson.Value)
              Just p ->
                if T.length p < 8
                  then do
                    status status400
                    json (Aeson.object ["error" .= ("Password too short" :: Text)] :: Aeson.Value)
                  else do
                    st <- liftIO $ readTVarIO state
                    if M.member u (usernameToId st)
                      then do
                        status status409
                        json (Aeson.object ["error" .= ("Username already exists" :: Text)] :: Aeson.Value)
                      else do
                        let newId = nextUserId st
                        let newUser = User newId u p
                        let newState = st
                              { nextUserId = newId + 1
                              , users = M.insert newId newUser (users st)
                              , usernameToId = M.insert u newId (usernameToId st)
                              }
                        liftIO $ atomically $ writeTVar state newState
                        status status201
                        json (Aeson.object ["id" .= newId, "username" .= u] :: Aeson.Value)

    post "/login" $ do
      req <- jsonData' :: ActionM LoginReq
      case (loginUsername req, loginPassword req) of
        (Just u, Just p) -> do
          st <- liftIO $ readTVarIO state
          case M.lookup u (usernameToId st) of
            Nothing -> do
              status status401
              json (Aeson.object ["error" .= ("Invalid credentials" :: Text)] :: Aeson.Value)
            Just uid ->
              case M.lookup uid (users st) of
                Nothing -> do
                  status status401
                  json (Aeson.object ["error" .= ("Invalid credentials" :: Text)] :: Aeson.Value)
                Just user ->
                  if password user == p
                    then do
                      token <- liftIO generateToken
                      let newState = st { sessions = M.insert token uid (sessions st) }
                      liftIO $ atomically $ writeTVar state newState
                      setHeader "Set-Cookie" $ TL.pack $ "session_id=" ++ T.unpack token ++ "; Path=/; HttpOnly"
                      status status200
                      json (Aeson.object ["id" .= userId user, "username" .= username user] :: Aeson.Value)
                    else do
                      status status401
                      json (Aeson.object ["error" .= ("Invalid credentials" :: Text)] :: Aeson.Value)
        _ -> do
          status status401
          json (Aeson.object ["error" .= ("Invalid credentials" :: Text)] :: Aeson.Value)

    post "/logout" $ do
      _ <- requireAuth state
      mSession <- getCookie "session_id"
      case mSession of
        Just token -> do
          st <- liftIO $ readTVarIO state
          let newState = st { sessions = M.delete token (sessions st) }
          liftIO $ atomically $ writeTVar state newState
        Nothing -> return ()
      status status200
      json (Aeson.object [] :: Aeson.Value)

    get "/me" $ do
      uid <- requireAuth state
      st <- liftIO $ readTVarIO state
      case M.lookup uid (users st) of
        Just user -> do
          status status200
          json (Aeson.object ["id" .= userId user, "username" .= username user] :: Aeson.Value)
        Nothing -> do
          status status500
          json (Aeson.object ["error" .= ("Internal error" :: Text)] :: Aeson.Value)

    put "/password" $ do
      uid <- requireAuth state
      req <- jsonData' :: ActionM PasswordReq
      st <- liftIO $ readTVarIO state
      case M.lookup uid (users st) of
        Just user ->
          case (oldPassword req, newPassword req) of
            (Just oldP, Just newP) ->
              if password user /= oldP
                then do
                  status status401
                  json (Aeson.object ["error" .= ("Invalid credentials" :: Text)] :: Aeson.Value)
                else if T.length newP < 8
                  then do
                    status status400
                    json (Aeson.object ["error" .= ("Password too short" :: Text)] :: Aeson.Value)
                  else do
                    let newUser = user { password = newP }
                    let newState = st { users = M.insert uid newUser (users st) }
                    liftIO $ atomically $ writeTVar state newState
                    status status200
                    json (Aeson.object [] :: Aeson.Value)
            (Just _, Nothing) -> do
              status status400
              json (Aeson.object ["error" .= ("Password too short" :: Text)] :: Aeson.Value)
            _ -> do
              status status401
              json (Aeson.object ["error" .= ("Invalid credentials" :: Text)] :: Aeson.Value)
        Nothing -> do
          status status500
          json (Aeson.object ["error" .= ("Internal error" :: Text)] :: Aeson.Value)

    get "/todos" $ do
      uid <- requireAuth state
      st <- liftIO $ readTVarIO state
      let userTodosList = [ t | t <- M.elems (todos st), todoUserId t == uid ]
      let sortedTodos = L.sortBy (compare `on` todoId) userTodosList
      status status200
      json sortedTodos

    post "/todos" $ do
      uid <- requireAuth state
      req <- jsonData' :: ActionM TodoCreateReq
      let mTitle = tcTitle req
      case mTitle of
        Nothing -> do
          status status400
          json (Aeson.object ["error" .= ("Title is required" :: Text)] :: Aeson.Value)
        Just t ->
          if T.null t
            then do
              status status400
              json (Aeson.object ["error" .= ("Title is required" :: Text)] :: Aeson.Value)
            else do
              now <- liftIO getCurrentTimeStr
              st <- liftIO $ readTVarIO state
              let newId = nextTodoId st
              let newTodo = Todo
                    { todoId = newId
                    , todoUserId = uid
                    , todoTitle = t
                    , todoDescription = fromMaybe "" (tcDescription req)
                    , todoCompleted = False
                    , todoCreatedAt = now
                    , todoUpdatedAt = now
                    }
              let newState = st
                    { nextTodoId = newId + 1
                    , todos = M.insert newId newTodo (todos st)
                    }
              liftIO $ atomically $ writeTVar state newState
              status status201
              json newTodo

    get "/todos/:id" $ do
      uid <- requireAuth state
      tidStr <- getPathParam
      case readMaybe (T.unpack tidStr) :: Maybe Int of
        Nothing -> do
          status status404
          json (Aeson.object ["error" .= ("Todo not found" :: Text)] :: Aeson.Value)
        Just tid -> do
          st <- liftIO $ readTVarIO state
          case M.lookup tid (todos st) of
            Just t | todoUserId t == uid -> do
              status status200
              json t
            _ -> do
              status status404
              json (Aeson.object ["error" .= ("Todo not found" :: Text)] :: Aeson.Value)

    put "/todos/:id" $ do
      uid <- requireAuth state
      tidStr <- getPathParam
      req <- jsonData' :: ActionM TodoUpdateReq
      case readMaybe (T.unpack tidStr) :: Maybe Int of
        Nothing -> do
          status status404
          json (Aeson.object ["error" .= ("Todo not found" :: Text)] :: Aeson.Value)
        Just tid -> do
          st <- liftIO $ readTVarIO state
          case M.lookup tid (todos st) of
            Just t | todoUserId t == uid -> do
              let mTitle = tuTitle req
              case mTitle of
                Just newT | T.null newT -> do
                  status status400
                  json (Aeson.object ["error" .= ("Title is required" :: Text)] :: Aeson.Value)
                _ -> do
                  now <- liftIO getCurrentTimeStr
                  let newTitle = fromMaybe (todoTitle t) (tuTitle req)
                  let newDesc = fromMaybe (todoDescription t) (tuDescription req)
                  let newComp = fromMaybe (todoCompleted t) (tuCompleted req)
                  let newTodo = t
                        { todoTitle = newTitle
                        , todoDescription = newDesc
                        , todoCompleted = newComp
                        , todoUpdatedAt = now
                        }
                  let newState = st { todos = M.insert tid newTodo (todos st) }
                  liftIO $ atomically $ writeTVar state newState
                  status status200
                  json newTodo
            _ -> do
              status status404
              json (Aeson.object ["error" .= ("Todo not found" :: Text)] :: Aeson.Value)

    delete "/todos/:id" $ do
      uid <- requireAuth state
      tidStr <- getPathParam
      case readMaybe (T.unpack tidStr) :: Maybe Int of
        Nothing -> do
          status status404
          json (Aeson.object ["error" .= ("Todo not found" :: Text)] :: Aeson.Value)
        Just tid -> do
          st <- liftIO $ readTVarIO state
          case M.lookup tid (todos st) of
            Just t | todoUserId t == uid -> do
              let newState = st { todos = M.delete tid (todos st) }
              liftIO $ atomically $ writeTVar state newState
              status status204
            _ -> do
              status status404
              json (Aeson.object ["error" .= ("Todo not found" :: Text)] :: Aeson.Value)

File: data_baseline/openrouter__qwen__qwen3.7-plus/rust/rep8/src/main.rs
use axum::{
    extract::{Path, State},
    http::{header::SET_COOKIE, HeaderMap, StatusCode},
    response::IntoResponse,
    routing::{delete, get, post, put},
    Json, Router,
};
use chrono::Utc;
use regex::Regex;
use serde::{Deserialize, Serialize};
use serde_json::Value;
use std::collections::HashMap;
use std::sync::{Arc, Mutex, RwLock};
use uuid::Uuid;

#[derive(Clone, Serialize, Deserialize)]
struct User {
    id: u32,
    username: String,
    #[serde(skip)]
    password: String,
}

#[derive(Clone, Serialize, Deserialize)]
struct Todo {
    id: u32,
    #[serde(skip)]
    user_id: u32,
    title: String,
    description: String,
    completed: bool,
    created_at: String,
    updated_at: String,
}

#[derive(Clone)]
struct AppState {
    users: Arc<RwLock<HashMap<u32, User>>>,
    next_user_id: Arc<Mutex<u32>>,
    todos: Arc<RwLock<HashMap<u32, Todo>>>,
    next_todo_id: Arc<Mutex<u32>>,
    sessions: Arc<RwLock<HashMap<String, u32>>>,
}

impl Default for AppState {
    fn default() -> Self {
        Self {
            users: Arc::new(RwLock::new(HashMap::new())),
            next_user_id: Arc::new(Mutex::new(1)),
            todos: Arc::new(RwLock::new(HashMap::new())),
            next_todo_id: Arc::new(Mutex::new(1)),
            sessions: Arc::new(RwLock::new(HashMap::new())),
        }
    }
}

fn get_session_id(headers: &HeaderMap) -> Option<String> {
    let cookie_header = headers.get("cookie")?.to_str().ok()?;
    let cookies: HashMap<&str, &str> = cookie_header
        .split(';')
        .filter_map(|c| {
            let mut parts = c.splitn(2, '=');
            Some((parts.next()?.trim(), parts.next()?.trim()))
        })
        .collect();
    cookies.get("session_id").map(|s| s.to_string())
}

fn get_session_user_id(headers: &HeaderMap, state: &AppState) -> Option<u32> {
    let sid = get_session_id(headers)?;
    state.sessions.read().unwrap().get(&sid).copied()
}

#[derive(Deserialize)]
struct RegisterRequest {
    username: Option<String>,
    password: Option<String>,
}

async fn register(
    State(state): State<AppState>,
    Json(req): Json<RegisterRequest>,
) -> impl IntoResponse {
    let username = match &req.username {
        Some(u) => u,
        None => {
            return (
                StatusCode::BAD_REQUEST,
                Json(serde_json::json!({"error": "Invalid username"})),
            )
        }
    };
    let password = match &req.password {
        Some(p) => p,
        None => {
            return (
                StatusCode::BAD_REQUEST,
                Json(serde_json::json!({"error": "Password too short"})),
            )
        }
    };

    if username.len() < 3 || username.len() > 50 {
        return (
            StatusCode::BAD_REQUEST,
            Json(serde_json::json!({"error": "Invalid username"})),
        );
    }
    let re = Regex::new(r"^[a-zA-Z0-9_]+$").unwrap();
    if !re.is_match(username) {
        return (
            StatusCode::BAD_REQUEST,
            Json(serde_json::json!({"error": "Invalid username"})),
        );
    }
    if password.len() < 8 {
        return (
            StatusCode::BAD_REQUEST,
            Json(serde_json::json!({"error": "Password too short"})),
        );
    }

    let mut users = state.users.write().unwrap();
    let exists = users.values().any(|u| u.username == *username);
    if exists {
        return (
            StatusCode::CONFLICT,
            Json(serde_json::json!({"error": "Username already exists"})),
        );
    }

    let new_id = {
        let mut next_id = state.next_user_id.lock().unwrap();
        let id = *next_id;
        *next_id += 1;
        id
    };

    let user = User {
        id: new_id,
        username: username.clone(),
        password: password.clone(),
    };
    users.insert(new_id, user.clone());

    (
        StatusCode::CREATED,
        Json(serde_json::json!({"id": new_id, "username": username})),
    )
}

#[derive(Deserialize)]
struct LoginRequest {
    username: String,
    password: String,
}

async fn login(
    State(state): State<AppState>,
    Json(req): Json<LoginRequest>,
) -> impl IntoResponse {
    let users = state.users.read().unwrap();
    let user = users
        .values()
        .find(|u| u.username == req.username && u.password == req.password);

    match user {
        Some(u) => {
            let session_id = Uuid::new_v4().to_string();
            state.sessions.write().unwrap().insert(session_id.clone(), u.id);

            let mut headers = HeaderMap::new();
            headers.insert(
                SET_COOKIE,
                format!("session_id={}; Path=/; HttpOnly", session_id)
                    .parse()
                    .unwrap(),
            );

            let body: Value = serde_json::json!({"id": u.id, "username": u.username});
            (StatusCode::OK, headers, Json(body)).into_response()
        }
        None => {
            let body: Value = serde_json::json!({"error": "Invalid credentials"});
            (StatusCode::UNAUTHORIZED, HeaderMap::new(), Json(body)).into_response()
        }
    }
}

async fn logout(
    headers: HeaderMap,
    State(state): State<AppState>,
) -> impl IntoResponse {
    if let Some(_user_id) = get_session_user_id(&headers, &state) {
        if let Some(sid) = get_session_id(&headers) {
            state.sessions.write().unwrap().remove(&sid);
        }
        let body: Value = serde_json::json!({});
        (StatusCode::OK, Json(body)).into_response()
    } else {
        let body: Value = serde_json::json!({"error": "Authentication required"});
        (StatusCode::UNAUTHORIZED, Json(body)).into_response()
    }
}

async fn me(
    headers: HeaderMap,
    State(state): State<AppState>,
) -> impl IntoResponse {
    if let Some(user_id) = get_session_user_id(&headers, &state) {
        let users = state.users.read().unwrap();
        if let Some(u) = users.get(&user_id) {
            let body: Value = serde_json::json!({"id": u.id, "username": u.username});
            (StatusCode::OK, Json(body)).into_response()
        } else {
            let body: Value = serde_json::json!({"error": "Authentication required"});
            (StatusCode::UNAUTHORIZED, Json(body)).into_response()
        }
    } else {
        let body: Value = serde_json::json!({"error": "Authentication required"});
        (StatusCode::UNAUTHORIZED, Json(body)).into_response()
    }
}

#[derive(Deserialize)]
struct PasswordRequest {
    old_password: Option<String>,
    new_password: Option<String>,
}

async fn update_password(
    headers: HeaderMap,
    State(state): State<AppState>,
    Json(req): Json<PasswordRequest>,
) -> impl IntoResponse {
    let user_id = match get_session_user_id(&headers, &state) {
        Some(id) => id,
        None => {
            let body: Value = serde_json::json!({"error": "Authentication required"});
            return (StatusCode::UNAUTHORIZED, Json(body)).into_response();
        }
    };

    let old_password = req.old_password.unwrap_or_default();
    let new_password = req.new_password.unwrap_or_default();

    let mut users = state.users.write().unwrap();
    if let Some(u) = users.get_mut(&user_id) {
        if u.password != old_password {
            let body: Value = serde_json::json!({"error": "Invalid credentials"});
            return (StatusCode::UNAUTHORIZED, Json(body)).into_response();
        }
        if new_password.len() < 8 {
            let body: Value = serde_json::json!({"error": "Password too short"});
            return (StatusCode::BAD_REQUEST, Json(body)).into_response();
        }
        u.password = new_password;
        let body: Value = serde_json::json!({});
        (StatusCode::OK, Json(body)).into_response()
    } else {
        let body: Value = serde_json::json!({"error": "Authentication required"});
        (StatusCode::UNAUTHORIZED, Json(body)).into_response()
    }
}

async fn list_todos(
    headers: HeaderMap,
    State(state): State<AppState>,
) -> impl IntoResponse {
    let user_id = match get_session_user_id(&headers, &state) {
        Some(id) => id,
        None => {
            let body: Value = serde_json::json!({"error": "Authentication required"});
            return (StatusCode::UNAUTHORIZED, Json(body)).into_response();
        }
    };

    let todos = state.todos.read().unwrap();
    let mut user_todos: Vec<_> = todos
        .values()
        .filter(|t| t.user_id == user_id)
        .cloned()
        .collect();

    user_todos.sort_by_key(|t| t.id);

    let body: Value = serde_json::to_value(user_todos).unwrap();
    (StatusCode::OK, Json(body)).into_response()
}

#[derive(Deserialize)]
struct CreateTodoRequest {
    title: Option<String>,
    description: Option<String>,
}

async fn create_todo(
    headers: HeaderMap,
    State(state): State<AppState>,
    Json(req): Json<CreateTodoRequest>,
) -> impl IntoResponse {
    let user_id = match get_session_user_id(&headers, &state) {
        Some(id) => id,
        None => {
            let body: Value = serde_json::json!({"error": "Authentication required"});
            return (StatusCode::UNAUTHORIZED, Json(body)).into_response();
        }
    };

    let title = match req.title {
        Some(t) if !t.is_empty() => t,
        _ => {
            let body: Value = serde_json::json!({"error": "Title is required"});
            return (StatusCode::BAD_REQUEST, Json(body)).into_response();
        }
    };

    let description = req.description.unwrap_or_default();
    let now = Utc::now().format("%Y-%m-%dT%H:%M:%SZ").to_string();

    let new_id = {
        let mut next_id = state.next_todo_id.lock().unwrap();
        let id = *next_id;
        *next_id += 1;
        id
    };

    let todo = Todo {
        id: new_id,
        user_id,
        title,
        description,
        completed: false,
        created_at: now.clone(),
        updated_at: now,
    };

    state.todos.write().unwrap().insert(new_id, todo.clone());

    let body: Value = serde_json::to_value(todo).unwrap();
    (StatusCode::CREATED, Json(body)).into_response()
}

async fn get_todo(
    headers: HeaderMap,
    State(state): State<AppState>,
    Path(todo_id): Path<u32>,
) -> impl IntoResponse {
    let user_id = match get_session_user_id(&headers, &state) {
        Some(id) => id,
        None => {
            let body: Value = serde_json::json!({"error": "Authentication required"});
            return (StatusCode::UNAUTHORIZED, Json(body)).into_response();
        }
    };

    let todos = state.todos.read().unwrap();
    if let Some(todo) = todos.get(&todo_id) {
        if todo.user_id == user_id {
            let body: Value = serde_json::to_value(todo.clone()).unwrap();
            (StatusCode::OK, Json(body)).into_response()
        } else {
            let body: Value = serde_json::json!({"error": "Todo not found"});
            (StatusCode::NOT_FOUND, Json(body)).into_response()
        }
    } else {
        let body: Value = serde_json::json!({"error": "Todo not found"});
        (StatusCode::NOT_FOUND, Json(body)).into_response()
    }
}

#[derive(Deserialize)]
struct UpdateTodoRequest {
    title: Option<String>,
    description: Option<String>,
    completed: Option<bool>,
}

async fn update_todo(
    headers: HeaderMap,
    State(state): State<AppState>,
    Path(todo_id): Path<u32>,
    Json(req): Json<UpdateTodoRequest>,
) -> impl IntoResponse {
    let user_id = match get_session_user_id(&headers, &state) {
        Some(id) => id,
        None => {
            let body: Value = serde_json::json!({"error": "Authentication required"});
            return (StatusCode::UNAUTHORIZED, Json(body)).into_response();
        }
    };

    let mut todos = state.todos.write().unwrap();
    if let Some(todo) = todos.get_mut(&todo_id) {
        if todo.user_id != user_id {
            let body: Value = serde_json::json!({"error": "Todo not found"});
            return (StatusCode::NOT_FOUND, Json(body)).into_response();
        }

        if let Some(title) = req.title {
            if title.is_empty() {
                let body: Value = serde_json::json!({"error": "Title is required"});
                return (StatusCode::BAD_REQUEST, Json(body)).into_response();
            }
            todo.title = title;
        }
        if let Some(desc) = req.description {
            todo.description = desc;
        }
        if let Some(completed) = req.completed {
            todo.completed = completed;
        }

        todo.updated_at = Utc::now().format("%Y-%m-%dT%H:%M:%SZ").to_string();

        let body: Value = serde_json::to_value(todo.clone()).unwrap();
        (StatusCode::OK, Json(body)).into_response()
    } else {
        let body: Value = serde_json::json!({"error": "Todo not found"});
        (StatusCode::NOT_FOUND, Json(body)).into_response()
    }
}

async fn delete_todo(
    headers: HeaderMap,
    State(state): State<AppState>,
    Path(todo_id): Path<u32>,
) -> impl IntoResponse {
    let user_id = match get_session_user_id(&headers, &state) {
        Some(id) => id,
        None => {
            let body: Value = serde_json::json!({"error": "Authentication required"});
            return (StatusCode::UNAUTHORIZED, Json(body)).into_response();
        }
    };

    let mut todos = state.todos.write().unwrap();
    if let Some(todo) = todos.get(&todo_id) {
        if todo.user_id == user_id {
            todos.remove(&todo_id);
            (StatusCode::NO_CONTENT, ()).into_response()
        } else {
            let body: Value = serde_json::json!({"error": "Todo not found"});
            (StatusCode::NOT_FOUND, Json(body)).into_response()
        }
    } else {
        let body: Value = serde_json::json!({"error": "Todo not found"});
        (StatusCode::NOT_FOUND, Json(body)).into_response()
    }
}

#[tokio::main]
async fn main() {
    let args: Vec<String> = std::env::args().collect();
    let mut port = 3000;
    for i in 0..args.len() {
        if args[i] == "--port" && i + 1 < args.len() {
            port = args[i + 1].parse().unwrap_or(3000);
        }
    }

    let state = AppState::default();

    let app = Router::new()
        .route("/register", post(register))
        .route("/login", post(login))
        .route("/logout", post(logout))
        .route("/me", get(me))
        .route("/password", put(update_password))
        .route("/todos", get(list_todos))
        .route("/todos", post(create_todo))
        .route("/todos/:id", get(get_todo))
        .route("/todos/:id", put(update_todo))
        .route("/todos/:id", delete(delete_todo))
        .with_state(state);

    let addr = format!("0.0.0.0:{}", port);
    println!("Listening on {}", addr);
    let listener = tokio::net::TcpListener::bind(addr).await.unwrap();
    axum::serve(listener, app).await.unwrap();
}
